lib/src/arithmetic/values/fixed_point_populator.dart

```diff
@@ -0,0 +1,136 @@
+  FpvType ofFixedPoint(FixedPoint fp) => populate(
+        integer: fp.integer.value,
+        fraction: fp.fraction.value,
+      );
+
+  @override
+  String toString() => 'FixedPointValuePopulator<${_unpopulated.runtimeType}>';
+
+  /// Populates the [FixedPointValue] with the given [integer] and
+  /// [fraction], then performs additional validation.
+  FpvType populate(
+      {required LogicValue integer, required LogicValue fraction}) {
+    if (_hasPopulated) {
+      throw RohdHclException('FixedPointPopulator: already populated');
+    }
+    _hasPopulated = true;
+
+    return _unpopulated
+      ..integer = integer
+      ..fraction = fraction;
+  }
+
+  /// Construct a [FixedPointValue] from a [LogicValue]
+  FpvType ofLogicValue(LogicValue val) => populate(
+        integer: val.getRange(
+            fractionWidth, integerWidth + fractionWidth + (signed ? 1 : 0)),
+        fraction: val.getRange(0, fractionWidth),
+      );
+
+  /// Return true if double [val] to be stored in [FixedPointValue]
+  /// with [m] and [n] lengths without overflowing.
+  static bool canStore(double val,
+      {required bool signed, required int m, required int n}) {
+    final w = signed ? 1 + m + n : m + n;
+    if (val.isFinite) {
+      final bigIntegerValue = BigInt.from(val * pow(2, n));
+      final negBigIntegerValue = BigInt.from(-val * pow(2, n));
+      final l = (val < 0.0)
+          ? max(bigIntegerValue.bitLength, negBigIntegerValue.bitLength)
+          : bigIntegerValue.bitLength;
+      return l <= w;
+    }
+    return false;
+  }
+
+  /// Constructs [FixedPointValue] from a Dart [double] rounding away from zero.
+  FpvType ofDouble(double val) {
+    final m = integerWidth;
+    final n = fractionWidth;
+    final signed = _unpopulated.signed;
+    if (!signed & (val < 0)) {
+      throw RohdHclException('Negative input not allowed with unsigned');
+    }
+    if (!canStore(val, signed: signed, m: m, n: n)) {
+      throw RohdHclException('Double is too long to store in '
+          'FixedPointValue: $m, $n');
+    }
+    final integerValue = BigInt.from(val * pow(2, n));
+    final w = signed ? 1 + m + n : m + n;
+    final v = LogicValue.ofBigInt(integerValue, w);
+    return ofLogicValue(v);
+  }
+
+  /// Constructs [FixedPointValue] from a Dart [double] without rounding.
+  FpvType ofDoubleUnrounded(double val) {
+    final m = integerWidth;
+    final n = fractionWidth;
+    final signed = _unpopulated.signed;
+    if (!signed & (val < 0)) {
+      throw RohdHclException('Negative input not allowed with unsigned');
+    }
+    final integerValue = BigInt.from(val * pow(2, n + 1));
+    final w = signed ? 1 + m + n : m + n;
+    final v = LogicValue.ofBigInt(integerValue >> 1, w);
+    return ofLogicValue(v);
+  }
+
+  /// Constructs a [FixedPointValue] from another [FixedPointValue] with
+  /// by widening the integer and/or fraction widths.
+  FpvType widen(FpvType fpv) {
```

Comment:
thought: what if we had a flag in `ofLogicValue` to allow widening rather than a separate function?

Reply:
I was thinking I should start naming variables with fxv instead of fpv to help when we work on conversion.

Reply:
> the issue is that this takes a FXV and needs to know what the original 'dot' location is to do the extension. Widening is relative to the 'dot', both directions.

ah, gotcha, thanks!

> I was thinking I should start naming variables with fxv instead of fpv to help when we work on conversion.

name change later isn't a big problem for positional arguments and type parameters i think, up to you when to do it if you want to